### DISCUSSION TITLE
Improve the handling of `sh.keptn.event.get-sli.triggered` events where SLIs don't produce a value

### DISCUSSION THREAD

**Discussion by arthurpitman** [keptn-contrib/dynatrace-service#704] [Ideas]
### Current situation
SLIs that produce no data during a given evaluation result in `sh.keptn.event.get-sli.finished` event with `result` set to `fail`. This in turn causes the evaluation to fail, as the lighthouse-service does not process the event further.

### Desired behavior
In cases where the dynatrace-service is able to query the Dyntrace tenant successfully, but the processing of the result leads to no single numerical value (i.e. when no data is available during the timeframe or no single value can be selected from the return results), `success` of the corresponding element in the `indicatorValues` is set to `false` but the overall `result` is set to `warning`.

This would mean the dynatrace-service would only function as an SLI provider, providing indicators with as much information as possible, making its behavior more obvious and easy to document. It would no longer perform any sort of evaluation, this would be done entirely by the lighthouse-service, depending on the scoring of the SLO configuration.

In short, if the user has set up the evaluation to tolerate a certain number of failing objectives, then it could still pass even when some SLIs don't produce a value. Other classes of errors would still be handled as before: situations where it was unable to process the request would lead to `sh.keptn.event.get-sli.finished` event  with `status` of `errored`, while situations where it is misconfigured (bad SLI syntax or errors from the Dynatrace tenant) would still result in a `result` set to `fail`.

The overall behavior can be summarized as follows:

| Case | Description | `status` | `result` | Consequence |
| --- | --- | --- | --- | --- |
| A | **The best case**<br>If the dynatrace-service was able to retrieve the metric from the Dynatrace API, i.e. the chain of calls succeeded (~HTTP 200-class) and process the result to produce a single value | `succeeded` | `pass` | - the `success` field of each indicator in `indicatorValues` is set to `true`<br> - an SLI retrieval was performed<br><br>-> good case (lighthouse-service can continue with evaluation task) |
| B | **The new case**<br>As above, but if the processing of the result cannot produce a single value<br>- due to no data points being returned, <br>- multiple data points being returned, <br>- or not being able to select a single result value (USQL queries). |  `succeeded` | `warning` | - the `success` field of the individual indicator in `indicatorValues` is set to `false`<br> - an SLI retrieval was performed<br><br> -> lighthouse-service has to deal with SLIs that have no value |
| C | **Errors, but the user should be able to fix them**<br>If an HTTP 400- class error occurs, due to e.g.:<br> - syntax errors in SLI files,<br> - dashboard syntax errors,<br> - or failure to retrieve the metric from the Dynatrace API | `succeeded` | `fail` | - the `success` field of each affected indicator in `indicatorValues` is set to `false`<br> - a partial SLI retrieval may have been performed but it should be disregarded<br><br> -> `get-sli.finished` terminates the sequence execution |
| D | **Something is seriously wrong**<br>If the dynatrace-service was unable to process the request (encounters HTTP 500-class errors) | `errored` | `fail` | the `success` field of all indicators in `indicatorValues` is set to `false`<br> - most likely no SLI retrieval was performed<br><br> -> `get-sli.finished` terminates the sequence execution |

**Comment by johannes-b** [comment]
@arthurpitman thanks for the summary and simplifying it further!
Looks good from my point of view.

**Reply by arthurpitman** [reply]
@johannes-b it would be good to be able to determine programmatically which indicator generated the warning. For this we would need to extend or change the `success` field of indicator. I think this would however require a change in the spec.

**Comment by agardnerIT** [comment]
As a starter, it would be nice to see at least which metrics failed. At the moment there is zero output.

If I have an SLO with `50` metrics and `1` fails, I get zero indicator results and absolutely no output in the bridge (or JSON) to show me where to start.

What would help tremendously is at least showing (somehow that `49` of my metrics were fine but this `1` failed and I need to remove that tile from the dashboard or remove that SLI from the SLO doc.

A customer I'm working with has exactly this issue which has lead to me having to create a utility which reads an `sli.yaml` file and `slo.yaml` file then directly calls the DT API (for all the SLIs in the `slo.yaml`) and print out to the console the names of the SLIs that produced no data. The customer can then investigate and remove from the `slo.yaml` file (or dashboard).
